### ISSUE TITLE
Enabling --signoff / -s on commits

### ISSUE THREAD

**Issue by jancborchardt** [desktop/desktop#5351]
**Please describe the problem you think should be solved**

A lot of open source projects work with the [Developer Certificate of Origin](https://developercertificate.org/) (like Linux, Nextcloud and others). Currently, commits created with GitHub Desktop can not be signed off and you would have to amend each one using the command line. Or you have to manually insert the `Signed-off-by: …` line in every commit.

- Documentation of --signoff: https://git-scm.com/docs/git-commit#git-commit--s
- More info on the DCO and why --signoff was added: https://stackoverflow.com/a/1962112
- Same issue exists in the Git &  GitHub integration of Atom: https://github.com/atom/github/issues/914
- Tangentially related to #78 (Setup gpg signing)

**Do you have any potential solutions in mind?**

Possible solutions are:

1. Simply sign off every commit automatically. The only thing to add to the commit message is:
```
Signed-off-by: Name <mail@example.com>
```
2. Or if automatic signing has any drawback, have a global (or per-project) setting on whether commits should be signed automatically.

**Comment by shiftkey** [comment]
@jancborchardt thanks for the suggestion!

A couple of questions I have about this:

 - Commits are always signed based on the `user.name` and `user.email` settings in Git. If we don't let the user choose this are they going to be confused?
 - When repositories require this, is the typical convention is for all commits to be signed? There's no real need to sign one commit but not the next other when working in a repository, right?

**Comment by perploug** [comment]
Personally I use a global git-hook to ensure that my commits in github desktop contains a sign-off automatically ([Sample docs](https://github.com/zalando/zalando.github.io/blob/9d7cbdf1e5e5b27838899ef685838460cdb69b6a/_docs/releasing/contributions.md#automatic-git-sign-off))

IMO having a global setting which enables gh desktop to automatically sign-off all commits would be very helpful. Having sign-off in a commit, even when not need, doesn't seem to have any drawbacks.

Bots like [DCO Bot ](https://github.com/probot/dco)requires all commits to have sign-off to pass the status check

**Comment by jancborchardt** [comment]
> Commits are always signed based on the user.name and user.email settings in Git. If we don't let the user choose this are they going to be confused?

A simple way this could be handled right now:
- If there is no user.name and user.email set, do not sign the commits.
- If they are set, don’t sign the commits "stealthily", but basically just prepopulate the the Description field.

In the future, there could be an interface to set these user.name and .email attributes (or is there already)?


> When repositories require this, is the typical convention is for all commits to be signed? There's no real need to sign one commit but not the next other when working in a repository, right?

Yes, usually that is the case. As @perploug mentioned there are bots which require all commits to have sign-off, and that’s the case for us too.


---------------------

And @shiftkey thanks a lot for your work bringing GitHub Desktop to Linux! Much appreciated! :tada:

**Comment by shiftkey** [comment]
@jancborchardt 

> In the future, there could be an interface to set these user.name and .email attributes (or is there already)?

It's available in the **File** | **Preferences** (macOS) or **File** | **Options** (Windows/Linux) menu:

<img width="596" src="https://user-images.githubusercontent.com/359239/43908837-58dc20fa-9bcf-11e8-93e9-4f2096ada814.png">

**Comment by jancborchardt** [comment]
@shiftkey right, sorry I should have checked. :) What do you think about the points about how automatic signing could be handled?

**Comment by billygriffin** [comment]
@jancborchardt Do you think the prepopulation in the description field would be a good idea across the board? I think it would have the potential to cause additional churn for folks on projects that don't require this sign off, as it's adding something into a text field that the user didn't actually type. Thoughts?

cc: @ampinsk

**Comment by jancborchardt** [comment]
@billygriffin no, I'd say that's more of a hack for the short term. Ideally you would have:

- A setting, either global or per-project, on if commits should be signed. Then if you don't need the sign-off, you won't see anything.
- If it will be signed, don't put anything specific in the commit view. Maybe only some kind of checkmark icon with a "Signed off " tooltip".

**Comment by perploug** [comment]
Agree with @jancborchardt: a small indicator like the +1 person icon at the bottom would be sufficient. 

One thing to consider tho, is that if you use different tools besides GH Desktop, you will have inconsistent commit signing, throughout my day I commit from desktop, cmdline and vscode - whatever is convenient. So consider if the solution would rather be to automate signing at a git-hook level, rather than on the tool level. GH Desktop could make this process easy by adding a global or project specific git-hook, which would then make signing consistent across whatever git tool you might use.

**Comment by billygriffin** [comment]
That's helpful, thanks for the context! @ampinsk Do you have thoughts on this? I'm not sure much user research would be that helpful here.

**Comment by jancborchardt** [comment]
@shiftkey @billygriffin @ampinsk is there any more info or designs needed? I’m also an interaction designer & user researcher so I’d love to help if that is what this is stuck at. :slightly_smiling_face:

**Comment by Naitik333** [comment]
@jancborchardt any updates on this? when can we expect automatic signoff in the commits through github desktop tool? currently after every commit we have to manually signoff them

**Comment by jancborchardt** [comment]
@Naitik333 I’m not working at GitHub so I can’t answer this, but I assume if there are any updates then this issue will be updated accordingly → so there must be no updates. :wink: 

It’s @shiftkey @billygriffin @ampinsk who would have to comment on specifics. I can also confirm this is still the single problem keeping Nextcloud contributors from using GitHub Desktop. :slightly_smiling_face:

**Comment by Naitik333** [comment]
@shiftkey @billygriffin @ampinsk  any updates? please confirm how you want to enable this feature so if you don't have time, than may be community can also help

**Comment by mislav** [comment]
@Naitik333 @jancborchardt The Desktop team does most of our work transparently, so if development of this feature had already started or if there were any other updates, we would have updated or referenced this thread.

Here is [our current roadmap](https://github.com/desktop/desktop/blob/development/docs/process/roadmap.md). Adding signoff to commits is not yet on it, and we don't know when and if we will prioritize it, so for now my general recommendation would be to make sure signoff is added outside of Desktop, preferrably via git hooks.

**Comment by visbits** [comment]
Bump.... this is needed.

**Comment by survivant** [comment]
I'm trying to fix my previous commit that I did from editing directly from github.com and I'm not able to add the signoff.  Why that feature is not available in github desktop ?   Could I just right click on a commit -> amend commit (signoff CHECKED).. and push.

**Comment by ImRodry** [comment]
This or an option to set a default commit description would be much appreciated as seen by the feedback on this issue. Bump

**Comment by ADustyOldMuffin** [comment]
I can also add that I'm running into issues with some repos when using Github Desktop to make commits. Looking into pre-hooks, but it would be nice if desktop could handle this instead.

**Comment by ADustyOldMuffin** [comment]
If there's a good design/idea setup I'd be happy to work on this.

**Comment by bveeramani** [comment]
Hey, just wanted to bump this issue. 

I prefer to use GitHub Desktop over Git in the command line, but I can't use GitHub Desktop because I need to sign off commits for work.

**Comment by sergiou87** [comment]
Another request for this in https://github.com/desktop/desktop/issues/19205

**Comment by hatamiarash7** [comment]
Any update about this issue?

**Comment by shitwolfymakes** [comment]
Bump: would appreciate this feature

**Comment by aakashagg** [comment]
Really need this feature.

**Comment by tedivm** [comment]
Kind of embarrassing that this feature hasn't been enabled yet.

**Comment by jancborchardt** [comment]
I'm as interested in this getting fixed as you recent commenters (which is why I opened the issue) but I can promise you that comments asking for an update or even negatively remarking on the project are of 0 help.

If you want to make your voice heard, upvote the original issue or become a contributor to the project. 🙏

**Comment by TheWitness** [comment]
Adding my $0.02 on this.... ppppplease, pretty pppppplease :).  Option below the 'Description' box.

**Comment by igalloway** [comment]
Bump: Would appreciate this feature. Recently started causing me problems.

FWIW as a user, I envision it as a checkbox on the commit dialog in the lower right of the desktop client.  Maybe an icon like the "add co-others" in that dialog.
I think literally this would this not just append the -s flag to the git background commit command if checked?
An alternative or additional option would be to add a checkbox in the <options><git> settings where your name and email and default branch are set.

**Comment by szaimen** [comment]
🎉🎉🎉🎉🎉

**Comment by TheWitness** [comment]
Nice!
